compiler/noirc_evaluator/src/ssa/opt/check_for_underconstrained_values.rs

```diff
@@ -0,0 +1,411 @@
+                if final_set.contains(&parameter_or_return_value) {
+                    connected_sets_indices.insert(set_index);
+                }
+            }
+        }
+        connected_sets_indices
+    }
+
+    /// Find which brillig calls separate this set from others and return bug warnings about them
+    fn find_disconnecting_brillig_calls_with_results_in_set(
+        &self,
+        current_set: &HashSet<ValueId>,
+        all_brillig_generated_values: &HashSet<ValueId>,
+        function: &Function,
+    ) -> Vec<SsaReport> {
+        let mut warnings = Vec::new();
+        // Find brillig-generated values in the set
+        let intersection = all_brillig_generated_values.intersection(current_set).copied();
+
+        // Go through all brillig outputs in the set
+        for brillig_output_in_set in intersection {
+            // Get the inputs that correspond to the output
+            let inputs: HashSet<ValueId> =
+                self.brillig_return_to_argument[&brillig_output_in_set].iter().copied().collect();
+
+            // Check if any of them are not in the set
+            let unused_inputs = inputs.difference(current_set).next().is_some();
+
+            // There is a value not in the set, which means that the inputs/outputs of this call have not been properly constrained
+            if unused_inputs {
+                warnings.push(SsaReport::Bug(InternalBug::IndependentSubgraph {
+                    call_stack: function.dfg.get_call_stack(
+                        self.brillig_return_to_instruction_id[&brillig_output_in_set],
+                    ),
+                }));
+            }
+        }
+        warnings
+    }
+    /// Go through each instruction in the block and add a set of ValueIds connected through that instruction
+    ///
+    /// Additionally, this function adds mappings of brillig return values to call arguments and instruction ids from calls to brillig functions in the block
+    fn connect_value_ids_in_block(
+        &mut self,
+        function: &Function,
+        block: BasicBlockId,
+        all_functions: &BTreeMap<FunctionId, Function>,
+    ) {
+        let instructions = function.dfg[block].instructions();
+
+        for instruction in instructions.iter() {
+            let mut instruction_arguments_and_results = HashSet::new();
+
+            // Insert non-constant instruction arguments
+            function.dfg[*instruction].for_each_value(|value_id| {
+                if function.dfg.get_numeric_constant(value_id).is_none() {
+                    instruction_arguments_and_results.insert(value_id);
+                }
+            });
+            // And non-constant results
+            for value_id in function.dfg.instruction_results(*instruction).iter() {
+                if function.dfg.get_numeric_constant(*value_id).is_none() {
+                    instruction_arguments_and_results.insert(*value_id);
+                }
+            }
```

Comment:
Is there an issue with unconditionally connecting results to their arguments here? If the results are from a brillig call they should not be connected, right?